db/refresh.go

```diff
@@ -188,6 +188,66 @@ func (r *refreshTokenRepo) Revoke(userID, token string) error {
 	return tx.Commit()
 }
 
+func (r *refreshTokenRepo) RenewRefreshToken(clientID, userID, oldToken string) (newRefreshToken string, err error) {
+	// Revoke old refresh token
+	tokenID, tokenPayload, err := parseToken(oldToken)
+	if err != nil {
+		return "", err
+	}
+
+	tx, err := r.begin()
+	if err != nil {
+		return "", err
+	}
+	defer tx.Rollback()
+	exec := r.executor(tx)
+	oldRecord, err := r.get(tx, tokenID)
+	if err != nil {
+		return "", err
+	}
+
+	if oldRecord.UserID != userID {
+		return "", refresh.ErrorInvalidUserID
+	}
+
+	if err := checkTokenPayload(oldRecord.PayloadHash, tokenPayload); err != nil {
+		return "", err
+	}
+
+	deleted, err := exec.Delete(oldRecord)
+	if err != nil {
+		return "", err
+	}
+	if deleted == 0 {
+		return "", refresh.ErrorInvalidToken
+	}
+
+	// Renew refresh token
+	tokenPayload, err = r.tokenGenerator.Generate()
+	if err != nil {
+		return "", err
+	}
+
+	payloadHash, err := bcrypt.GenerateFromPassword(tokenPayload, bcrypt.DefaultCost)
```

Comment:
I'm going to add a comment. It checks if the hash of payload from received refresh token are equal to the stored in database.